### PR TITLE
Added missing aspects of Realm.swift

### PR DESCRIPTION
**Pull request by jpsim**

### CODE REVIEW

**Comment by segiddins** [comment]
This looks pretty good to me

**Comment by segiddins** [comment]
@tgoyne @alazier I think this might be ready for review?

**Comment by alazier** [comment]
This is failing on CI - also seems to need a rebase/merge

**Comment by jpsim** [comment]
I like the changes you made in here :smile:

File: RealmSwift/Realm.swift [hunk 1]
```diff
@@ -98,73 +55,35 @@ public class Realm {
...
     // MARK: Initializers
 
-    init(rlmRealm: RLMRealm) {
-        self.rlmRealm = rlmRealm
-    }
-
     /**
-    Obtains a Realm instance persisted at the specified file path.
+    Obtains a Realm instance persisted at the specified file path. The default
+    Realm path is used by default.
```
Comment (tgoyne) [inline RealmSwift/Realm.swift]: This is a dumb sentence (i.e. find some other phrasing for "by default").

File: RealmSwift/Realm.swift [hunk 2]
```diff
@@ -302,43 +177,38 @@ public class Realm {
...
-    }
+    Revert all writes made in the current write transaction and end the transaction.
 
-    // MARK: Invalidation
+    This rolls back all objects in the Realm to the state they were in at the
+    beginning of the write transaction, and then ends the transaction.
 
-    /**
-    Invalidate all `Object`s and `Results` read from this Realm.
+    This restores the data for deleted objects, but does not re-validated deleted
```
Comment (tgoyne) [inline RealmSwift/Realm.swift]: re-validate

obj-c docs probably have the same typo

Comment (tgoyne) [inline RealmSwift/Realm.swift]: re-validate doesn't even mean that though

Comment (alazier) [inline RealmSwift/Realm.swift]: Would `reinstate` be a better term?

File: RealmSwift/Realm.swift [hunk 3]
```diff
@@ -435,8 +296,10 @@ public class Realm {
 
     :param: object The objects to be deleted.
     */
-    public func delete(objects: Results<Object>) {
-        rlmRealm.deleteObjects(objects)
+    public func delete<S: SequenceType where S.Generator.Element: Object>(objects: S) {
```
Comment (tgoyne) [inline RealmSwift/Realm.swift]: The Results overload does need to exist. rlmRealm.deleteObjects(objects) does something very different from looping over the items.

File: RealmSwift/Realm.swift [hunk 4]
```diff
@@ -475,18 +352,144 @@ public class Realm {
...
-    // MARK: Object Retrieval
+
+    // MARK: Autorefresh and Refresh
 
     /**
-    Returns all objects of the given type in this Realm.
+    Whether this Realm automatically updates when changes happen in other threads.
 
-    :param: type The type of the objects to be returned.
+    If set to YES (the default), changes made on other threads will be reflected
```
Comment (tgoyne) [inline RealmSwift/Realm.swift]: true/false rather than YES/NO for Swift.

File: RealmSwift/Realm.swift [hunk 5]
```diff
@@ -475,18 +352,144 @@ public class Realm {
...
+    committed.  If set to NO, you must manually call -refresh on the Realm to
+    update it to get the latest version.
+
+    Even with this enabled, you can still call `refresh()` at any time to update the
+    Realm before the automatic refresh would occur.
+
+    Notifications are sent when a write transaction is committed whether or not
+    this is enabled.
+
+    Disabling this on an `Realm` without any strong references to it will not
```
Comment (tgoyne) [inline RealmSwift/Realm.swift]: a Realm

File: RealmSwift/Tests/ListTests.swift [hunk 6]
```diff
@@ -267,7 +267,7 @@ class ListTests: TestCase {
     }
 
     func testAppendResults() {
-        array.append(realmWithTestPath().objects(SwiftStringObject))
+        array.append(realmWithTestPath().objects(SwiftStringObject.self))
```
Comment (tgoyne) [inline RealmSwift/Tests/ListTests.swift]: are these .self actually needed?

**Comment by tgoyne** [comment]
obj-c changes should also be cherry-picked to a separate branch and merged to master.

File: RealmSwift/Realm.swift (hunk 3 above)
Comment (alazier) [inline RealmSwift/Realm.swift]: That old results method did not work at all (nor does the current on which takes List). They need to be: 
```
    public func delete<T: Object>(objects: List<T>) {
        rlmRealm.deleteObjects(objects._rlmArray)
    }

    public func delete<T: Object>(objects: Results<T>) {
        rlmRealm.deleteObjects(objects.rlmResults)
    }
```

File: RealmSwift/Tests/ListTests.swift (hunk 6 above)
Comment (alazier) [inline RealmSwift/Tests/ListTests.swift]: They are no longer required (thank god)

**Comment by alazier** [comment]
@tgoyne @segiddins please take a quick look at the fixes before I merge.

**Comment by alazier** [comment]
@tgoyne cherry-picking the one change against master doesn't work because there are changes to swift files as well. @segiddins we should just make a new pr with these changes on master. Can comment out the two failing swift tests until this is done.

File: RealmSwift/Tests/RealmTests.swift [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+
+    func testInit() {
+        XCTAssertEqual(Realm().path, Realm.defaultPath)
+        XCTAssertEqual(Realm(path: testRealmPath()).path, testRealmPath())
+        assertThrows(Realm(path: ""))
+    }
+
+    func testInitFailable() {
+        var error: NSError?
+        autoreleasepool({
```
Comment (tgoyne) [inline RealmSwift/Tests/RealmTests.swift]: I think we've been consistent about dropping the parens for `autoreleasepool`. There's also a few `realm.write`s below with extra parens.

File: RealmSwift/Tests/RealmTests.swift [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+
+    func testSetEncryptionKey() {
+        Realm.setEncryptionKey(NSMutableData(length: 64))
+        Realm.setEncryptionKey(nil, forPath: Realm.defaultPath)
+        XCTAssert(true, "setting those keys should not throw")
+    }
+
+    func testEquals() {
+        let realm = Realm()
+        XCTAssertTrue(realm == Realm())
```
Comment (tgoyne) [inline RealmSwift/Tests/RealmTests.swift]: Does `XCTAssertEqual` not work here, or are you just being explicit about it testing `==`?

Comment (segiddins) [inline RealmSwift/Tests/RealmTests.swift]: Yes, being explicit about testing `==`, since that's the function implemented

**Pull request merged**